hashintel/hash
# Upgrade HASH to use Block Protocol 0.2

## 🌟 What is the purpose of this PR?

Upgrade HASH to use Block Protocol 0.2

The bulk of the changes are caused by the need to:
1. Updating the function signatures and calls, as per [the new graph specification](https://blockprotocol.org/docs/spec/graph-service-specification) (all the `hooks/blockProtocolFunctions` files)
2. Rewrite the ids and shapes for entities and links.

There are a few more changes detailed below.

This PR depends on a new version of `@blockprotocol/graph` published from https://github.com/blockprotocol/blockprotocol/pull/421 – once both PRs have been reviewed & tested I will cut a further, stable/latest release of `graph` and update the version numbers.

## 🔗 Related links

- [Asana task for updating to 0.2](https://app.asana.com/0/1200211978612931/1202353766855920/f) _(internal)_
- [Task for sending entity types to blocks](https://app.asana.com/0/1200211978612931/1202229645898395/f) _(internal)_

## 🔍 What does this change?

- Rewrite identifiers for entities and links when sending them to blocks: at the moment we need to send `accountId` to our API when identifying entities and links, so this PR introduces rewriting of the `entityId` (and similar) that is sent to blocks so that it contains the accountId (which is then parsed again when the BP functions are called).
- Update shape for entities and types send to blocks: these now have a nested `properties` (or `schema`) object rather than their own properties appearing alongside BP-specified identifiers (`entityId` or `entityTypeId`)
- Starts passing all the `entityTypes` for the entities send to blocks (I've left a [follow-up](https://app.asana.com/0/1202482456346915/1202566631614518/f) to feed through the block entity entity type, which requires a few more changes)
- Updates the HTML block renderer, adds a custom element block renderer
- Update imports for a few utility types from `blockprotocol` to `@blockprotocol/core`
- Fix the entity editor and entity type editor for the new approach
- Remove a couple of deprecated things:
  - `metadataId`: this was an old alias for `entityId` that was only being used in one place, where the page title was being updated.
  - `[blockVariant].displayName`
- comments out some sandbox code, which will be updated for 0.2 as part of [this task](https://app.asana.com/0/1200211978612931/1201901731799594/f)

## 🐾 Next steps

 1. Once this and https://github.com/blockprotocol/blockprotocol/pull/421 have been reviewed & tested I will cut a further, stable/latest release of `graph` and update the version numbers.
 2. Update BP docs to mention HASH as an option for testing blocks
 3. Fix sandboxing in HASH
 4. Eventually fully remove the deprecated `blockprotocol` package from the repo (a few blocks and the WIP new entity editor depend on it)

## 🛡 What tests cover this?

- Existing tests

## ❓ How to test this?

1.  View / create a page with different blocks, check they work
5. Try the entity type editor
6. Try the entity editor (select a type from the sidebar and click 'new [thing]'
